File: react-native-nitro-player/android/src/main/java/com/margelo/nitro/nitroplayer/HybridTrackPlayer.kt
@file:Suppress("ktlint:standard:max-line-length")

package com.margelo.nitro.nitroplayer

import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import com.margelo.nitro.nitroplayer.core.TrackPlayerCore
import com.margelo.nitro.nitroplayer.core.addToUpNextInternal
import com.margelo.nitro.nitroplayer.core.clearPlayNextOnQueue
import com.margelo.nitro.nitroplayer.core.clearUpNextOnQueue
import com.margelo.nitro.nitroplayer.core.configureOnQueue
import com.margelo.nitro.nitroplayer.core.getActualQueueInternal
import com.margelo.nitro.nitroplayer.core.getNextTracksInternal
import com.margelo.nitro.nitroplayer.core.getRepeatMode
import com.margelo.nitro.nitroplayer.core.getStateInternal
import com.margelo.nitro.nitroplayer.core.getTracksNeedingUrlsInternal
import com.margelo.nitro.nitroplayer.core.pauseOnQueue
import com.margelo.nitro.nitroplayer.core.playNextInternal
import com.margelo.nitro.nitroplayer.core.playOnQueue
import com.margelo.nitro.nitroplayer.core.playSongInternal
import com.margelo.nitro.nitroplayer.core.removeFromPlayNextOnQueue
import com.margelo.nitro.nitroplayer.core.removeFromUpNextOnQueue
import com.margelo.nitro.nitroplayer.core.reorderTemporaryTrackOnQueue
import com.margelo.nitro.nitroplayer.core.seekOnQueue
import com.margelo.nitro.nitroplayer.core.setPlayBackSpeedOnQueue
import com.margelo.nitro.nitroplayer.core.setRepeatModeOnQueue
import com.margelo.nitro.nitroplayer.core.setVolumeOnQueue
import com.margelo.nitro.nitroplayer.core.skipToIndexInternal
import com.margelo.nitro.nitroplayer.core.skipToNextOnQueue
import com.margelo.nitro.nitroplayer.core.skipToPreviousOnQueue
import com.margelo.nitro.nitroplayer.core.updateTracksOnQueue

@DoNotStrip
@Keep
class HybridTrackPlayer : HybridTrackPlayerSpec() {
    private val core: TrackPlayerCore

    // Track listener IDs for cleanup in dispose()
    private val listenerIds = mutableListOf<Pair<String, Long>>()

    init {
        val context =
            NitroModules.applicationContext
                ?: throw IllegalStateException("React Context is not initialized")
        core = TrackPlayerCore.getInstance(context)
    }

    // ── Ordered dispatch ──────────────────────────────────────────────────────
    //
    // `Promise.async { … }` launches an unordered coroutine, so the hop onto the player
    // looper happened at an arbitrary later point: two calls made in order from JS could
    // reach the player in reverse order (play-then-pause landing as pause-then-play).
    // `core.enqueue` appends synchronously, on the JS thread at call time, so execution
    // order equals JS call order.

    private fun <T> enqueue(block: () -> T): Promise<T> {
        val promise = Promise<T>()
        core.enqueue {
            try {
                promise.resolve(block())
            } catch (e: Throwable) {
                promise.reject(e)
            }
        }
        return promise
    }

    // ── Playback ─────────────────────────────────────────────────────────────

    override fun play(): Promise<Unit> = enqueue { core.playOnQueue() }

    override fun pause(): Promise<Unit> = enqueue { core.pauseOnQueue() }

    override fun seek(position: Double): Promise<Unit> = enqueue { core.seekOnQueue(position) }

    override fun skipToNext(): Promise<Unit> = enqueue { core.skipToNextOnQueue() }

    override fun skipToPrevious(): Promise<Unit> = enqueue { core.skipToPreviousOnQueue() }

    override fun playSong(
        songId: String,
        fromPlaylist: String?,
    ): Promise<Unit> = enqueue { core.playSongInternal(songId, fromPlaylist) }

    override fun skipToIndex(index: Double): Promise<Boolean> = enqueue { core.skipToIndexInternal(index.toInt()) }

    override fun setRepeatMode(mode: RepeatMode): Promise<Unit> = enqueue { core.setRepeatModeOnQueue(mode) }

    override fun getRepeatMode(): RepeatMode = core.getRepeatMode()

    override fun setVolume(volume: Double): Promise<Unit> = enqueue { core.setVolumeOnQueue(volume) }

    override fun configure(config: PlayerConfig): Promise<Unit> = enqueue { core.configureOnQueue(config) }

    // ── Queue / state reads ───────────────────────────────────────────────────

    override fun getActualQueue(): Promise<Array<TrackItem>> = enqueue { core.getActualQueueInternal().toTypedArray() }

    override fun getState(): Promise<PlayerState> = enqueue { core.getStateInternal() }

    override fun getTracksById(trackIds: Array<String>): Promise<Array<TrackItem>> = enqueue { core.getPlaylistManager().getTracksById(trackIds.toList()).toTypedArray() }

    override fun getTracksNeedingUrls(): Promise<Array<TrackItem>> = enqueue { core.getTracksNeedingUrlsInternal().toTypedArray() }

    override fun getNextTracks(count: Double): Promise<Array<TrackItem>> = enqueue { core.getNextTracksInternal(count.toInt()).toTypedArray() }

    override fun getCurrentTrackIndex(): Promise<Double> = enqueue { core.currentTrackIndex.toDouble() }

    // ── URL updates ───────────────────────────────────────────────────────────

    override fun updateTracks(tracks: Array<TrackItem>): Promise<Unit> = enqueue { core.updateTracksOnQueue(tracks.toList()) }

    // ── Temporary queue ───────────────────────────────────────────────────────

    override fun addToUpNext(trackId: String): Promise<Unit> = enqueue { core.addToUpNextInternal(trackId) }

    override fun playNext(trackId: String): Promise<Unit> = enqueue { core.playNextInternal(trackId) }

    override fun removeFromPlayNext(trackId: String): Promise<Boolean> = enqueue { core.removeFromPlayNextOnQueue(trackId) }

    override fun removeFromUpNext(trackId: String): Promise<Boolean> = enqueue { core.removeFromUpNextOnQueue(trackId) }

    override fun clearPlayNext(): Promise<Unit> = enqueue { core.clearPlayNextOnQueue() }

    override fun clearUpNext(): Promise<Unit> = enqueue { core.clearUpNextOnQueue() }

    override fun reorderTemporaryTrack(
        trackId: String,
        newIndex: Double,
    ): Promise<Boolean> = enqueue { core.reorderTemporaryTrackOnQueue(trackId, newIndex.toInt()) }

    override fun getPlayNextQueue(): Promise<Array<TrackItem>> = enqueue { core.playNextStack.toTypedArray() }

    override fun getUpNextQueue(): Promise<Array<TrackItem>> = enqueue { core.upNextQueue.toTypedArray() }

    // ── Playback speed ────────────────────────────────────────────────────────

    override fun setPlaybackSpeed(speed: Double): Promise<Unit> = enqueue { core.setPlayBackSpeedOnQueue(speed) }

    override fun getPlaybackSpeed(): Promise<Double> = enqueue { if (core.isExoInitialized) core.exo.getPlaybackSpeed().toDouble() else 1.0 }

    // ── Android Auto ──────────────────────────────────────────────────────────

    override fun isAndroidAutoConnected(): Boolean = core.isAndroidAutoConnected()

    // ── Event listeners ───────────────────────────────────────────────────────

    override fun onChangeTrack(callback: (track: TrackItem, reason: Reason?) -> Unit) {
        val id = core.addOnChangeTrackListener(callback)
        listenerIds += "onChangeTrack" to id
    }

    override fun onPlaybackStateChange(callback: (state: TrackPlayerState, reason: Reason?) -> Unit) {
        val id = core.addOnPlaybackStateChangeListener(callback)
        listenerIds += "onPlaybackStateChange" to id
    }

    override fun onSeek(callback: (position: Double, totalDuration: Double) -> Unit) {
        val id = core.addOnSeekListener(callback)
        listenerIds += "onSeek" to id
    }

    override fun onPlaybackProgressChange(callback: (position: Double, totalDuration: Double, isManuallySeeked: Boolean?) -> Unit) {
        val id = core.addOnPlaybackProgressChangeListener(callback)
        listenerIds += "onPlaybackProgressChange" to id
    }

    override fun onAndroidAutoConnectionChange(callback: (connected: Boolean) -> Unit) {
        // Replace any prior listener registered through this HybridObject so JS
        // reloads (Fast Refresh, hot reload) don't accumulate handlers.
        val existing = listenerIds.filter { it.first == "onAndroidAutoConnectionChange" }
        existing.forEach { (_, oldId) -> core.removeOnAndroidAutoConnectionListener(oldId) }
        listenerIds.removeAll { it.first == "onAndroidAutoConnectionChange" }

        val id = core.addOnAndroidAutoConnectionListener(callback)
        listenerIds += "onAndroidAutoConnectionChange" to id
    }

    override fun onTracksNeedUpdate(callback: (tracks: Array<TrackItem>, lookahead: Double) -> Unit) {
        val id =
            core.addOnTracksNeedUpdateListener { tracks, lookahead ->
                callback(tracks.toTypedArray(), lookahead.toDouble())
            }
        listenerIds += "onTracksNeedUpdate" to id
    }

    override fun onTimedMetadata(callback: (metadata: TimedMetadata) -> Unit) {
        val id = core.addOnTimedMetadataListener(callback)
        listenerIds += "onTimedMetadata" to id
    }

    override fun onTemporaryQueueChange(callback: (playNextQueue: Array<TrackItem>, upNextQueue: Array<TrackItem>) -> Unit) {
        val id =
            core.addOnTemporaryQueueChangeListener { pn, un ->
                callback(pn.toTypedArray(), un.toTypedArray())
            }
        listenerIds += "onTemporaryQueueChange" to id
    }

    // ── Cleanup ───────────────────────────────────────────────────────────────

    override fun dispose() {
        super.dispose()
        listenerIds.forEach { (type, id) ->
            when (type) {
                "onChangeTrack" -> core.removeOnChangeTrackListener(id)
                "onPlaybackStateChange" -> core.removeOnPlaybackStateChangeListener(id)
                "onSeek" -> core.removeOnSeekListener(id)
                "onPlaybackProgressChange" -> core.removeOnPlaybackProgressChangeListener(id)
                "onAndroidAutoConnectionChange" -> core.removeOnAndroidAutoConnectionListener(id)
                "onTracksNeedUpdate" -> core.removeOnTracksNeedUpdateListener(id)
                "onTemporaryQueueChange" -> core.removeOnTemporaryQueueChangeListener(id)
                "onTimedMetadata" -> core.removeOnTimedMetadataListener(id)
            }
        }
        listenerIds.clear()
    }
}

File: react-native-nitro-player/ios/HybridTrackPlayer.swift
//
//  HybridTrackPlayer.swift
//  NitroPlayer
//
//  Created by Ritesh Shukla on 10/12/25.
//

import Foundation
import NitroModules

/// Hybrid implementation of TrackPlayerSpec for iOS
/// Bridges Nitro modules with the native TrackPlayerCore implementation
final class HybridTrackPlayer: HybridTrackPlayerSpec {
  // MARK: - Properties

  private let core: TrackPlayerCore

  /// Stable listener IDs for cleanup on deinit
  private var listenerIds: [(String, Int64)] = []

  // MARK: - Initialization

  override init() {
    core = TrackPlayerCore.shared
    super.init()
  }

  // MARK: - Ordered dispatch
  //
  // `Promise.async` starts an unordered Task, so the hop onto the player queue used
  // to happen at an arbitrary later point: two calls made in order from JS could
  // reach the player in reverse order (play-then-pause landing as pause-then-play).
  // Enqueueing synchronously here — on the JS thread, at call time — makes the serial
  // queue's FIFO order equal the JS call order.

  private func enqueue<T>(_ block: @escaping () -> T) -> Promise<T> {
    let promise = Promise<T>()
    core.playerQueue.async { promise.resolve(withResult: block()) }
    return promise
  }

  private func enqueueThrowing<T>(_ block: @escaping () throws -> T) -> Promise<T> {
    let promise = Promise<T>()
    core.playerQueue.async {
      do { promise.resolve(withResult: try block()) }
      catch { promise.reject(withError: error) }
    }
    return promise
  }

  // MARK: - Playback Control (async Promise<Void>)

  func play() throws -> Promise<Void> {
    enqueue { self.core.playOnQueue() }
  }

  func pause() throws -> Promise<Void> {
    enqueue { self.core.pauseOnQueue() }
  }

  func seek(position: Double) throws -> Promise<Void> {
    enqueue { self.core.seekOnQueue(position: position) }
  }

  func skipToNext() throws -> Promise<Void> {
    enqueue { self.core.skipToNextOnQueue() }
  }

  func skipToPrevious() throws -> Promise<Void> {
    enqueue { self.core.skipToPreviousOnQueue() }
  }

  func playSong(songId: String, fromPlaylist: String?) throws -> Promise<Void> {
    enqueue { self.core.playSongInternal(songId: songId, fromPlaylist: fromPlaylist) }
  }

  func skipToIndex(index: Double) throws -> Promise<Bool> {
    enqueue { self.core.skipToIndexOnQueue(index: Int(index)) }
  }

  // MARK: - Repeat / Volume / Config

  func setRepeatMode(mode: RepeatMode) throws -> Promise<Void> {
    enqueue { self.core.setRepeatModeOnQueue(mode: mode) }
  }

  func getRepeatMode() throws -> RepeatMode {
    core.getRepeatMode()
  }

  func setVolume(volume: Double) throws -> Promise<Void> {
    enqueue { self.core.setVolumeOnQueue(volume: volume) }
  }

  func configure(config: PlayerConfig) throws -> Promise<Void> {
    enqueue {
      self.core.configureOnQueue(
        androidAutoEnabled: config.androidAutoEnabled,
        carPlayEnabled: config.carPlayEnabled,
        showInNotification: config.showInNotification,
        lookaheadCount: config.lookaheadCount.map { Int($0) }
      )
    }
  }

  // MARK: - Queue / State reads

  func getActualQueue() throws -> Promise<[TrackItem]> {
    enqueue { self.core.getActualQueueInternal() }
  }

  func getState() throws -> Promise<PlayerState> {
    enqueue { self.core.getStateInternal() }
  }

  func getCurrentTrackIndex() throws -> Promise<Double> {
    enqueue { Double(self.core.currentTrackIndex) }
  }

  // MARK: - URL updates / lazy loading

  func updateTracks(tracks: [TrackItem]) throws -> Promise<Void> {
    enqueue { self.core.updateTracksInternal(tracks: tracks) }
  }

  func getTracksById(trackIds: [String]) throws -> Promise<[TrackItem]> {
    enqueue { self.core.getPlaylistManager().getTracksById(trackIds: trackIds) }
  }

  func getTracksNeedingUrls() throws -> Promise<[TrackItem]> {
    enqueue { self.core.getTracksNeedingUrlsInternal() }
  }

  func getNextTracks(count: Double) throws -> Promise<[TrackItem]> {
    enqueue { self.core.getNextTracksInternal(count: Int(count)) }
  }

  // MARK: - Playback speed
  func setPlaybackSpeed(speed: Double) throws -> Promise<Void> {
    enqueue { self.core.setPlaybackSpeedOnQueue(speed) }
  }

  func getPlaybackSpeed() throws -> Promise<Double> {
    enqueue { self.core.currentPlaybackSpeed }
  }

  // MARK: - Temporary queue v2

  func addToUpNext(trackId: String) throws -> Promise<Void> {
    enqueueThrowing { try self.core.addToUpNextOnQueue(trackId: trackId) }
  }

  func playNext(trackId: String) throws -> Promise<Void> {
    enqueueThrowing { try self.core.playNextOnQueue(trackId: trackId) }
  }

  func removeFromPlayNext(trackId: String) throws -> Promise<Bool> {
    enqueue { self.core.removeFromPlayNextOnQueue(trackId: trackId) }
  }

  func removeFromUpNext(trackId: String) throws -> Promise<Bool> {
    enqueue { self.core.removeFromUpNextOnQueue(trackId: trackId) }
  }

  func clearPlayNext() throws -> Promise<Void> {
    enqueue { self.core.clearPlayNextOnQueue() }
  }

  func clearUpNext() throws -> Promise<Void> {
    enqueue { self.core.clearUpNextOnQueue() }
  }

  func reorderTemporaryTrack(trackId: String, newIndex: Double) throws -> Promise<Bool> {
    enqueue { self.core.reorderTemporaryTrackOnQueue(trackId: trackId, newIndex: Int(newIndex)) }
  }

  func getPlayNextQueue() throws -> Promise<[TrackItem]> {
    enqueue { self.core.playNextStack }
  }

  func getUpNextQueue() throws -> Promise<[TrackItem]> {
    enqueue { self.core.upNextQueue }
  }

  // MARK: - Android Auto (iOS no-op)

  func onAndroidAutoConnectionChange(callback: @escaping (Bool) -> Void) throws {
    // No-op on iOS
  }

  func isAndroidAutoConnected() throws -> Bool { false }

  // MARK: - Event listeners (v2 — store IDs for cleanup)

  func onChangeTrack(callback: @escaping (_ track: TrackItem, _ reason: Reason?) -> Void) throws {
    let id = core.addOnChangeTrackListener(callback)
    listenerIds.append(("onChangeTrack", id))
  }

  func onPlaybackStateChange(callback: @escaping (_ state: TrackPlayerState, _ reason: Reason?) -> Void) throws {
    let id = core.addOnPlaybackStateChangeListener(callback)
    listenerIds.append(("onPlaybackStateChange", id))
  }

  func onSeek(callback: @escaping (_ position: Double, _ totalDuration: Double) -> Void) throws {
    let id = core.addOnSeekListener(callback)
    listenerIds.append(("onSeek", id))
  }

  func onPlaybackProgressChange(callback: @escaping (_ position: Double, _ totalDuration: Double, _ isManuallySeeked: Bool?) -> Void) throws {
    let id = core.addOnProgressListener(callback)
    listenerIds.append(("onPlaybackProgressChange", id))
  }

  func onTracksNeedUpdate(callback: @escaping (_ tracks: [TrackItem], _ lookahead: Double) -> Void) throws {
    let id = core.addOnTracksNeedUpdateListener { tracks, lookahead in
      callback(tracks, Double(lookahead))
    }
    listenerIds.append(("onTracksNeedUpdate", id))
  }

  func onTemporaryQueueChange(callback: @escaping (_ playNextQueue: [TrackItem], _ upNextQueue: [TrackItem]) -> Void) throws {
    let id = core.addOnTemporaryQueueChangeListener(callback)
    listenerIds.append(("onTemporaryQueueChange", id))
  }

  // MARK: - Cleanup

  deinit {
    for (type, id) in listenerIds {
      switch type {
      case "onChangeTrack":           _ = core.removeOnChangeTrackListener(id: id)
      case "onPlaybackStateChange":   _ = core.removeOnPlaybackStateChangeListener(id: id)
      case "onSeek":                  _ = core.removeOnSeekListener(id: id)
      case "onPlaybackProgressChange":_ = core.removeOnProgressListener(id: id)
      case "onTracksNeedUpdate":      _ = core.removeOnTracksNeedUpdateListener(id: id)
      case "onTemporaryQueueChange":  _ = core.removeOnTemporaryQueueChangeListener(id: id)
      default: break
      }
    }
  }
}
